src/engine/strat_engine/ns.rs

```diff
@@ -0,0 +1,90 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::{
+    fs::create_dir_all,
+    io,
+    path::{Path, PathBuf},
+};
+
+use nix::{
+    mount::{mount, umount, MsFlags},
+    sys::stat::stat,
+    unistd::gettid,
+};
+
+use crate::stratis::{StratisError, StratisResult};
+
+/// Path to the root mount namespace
+const INIT_MNT_NS_PATH: &str = "/proc/1/ns/mnt";
+/// Path to where private namespace mounts are mounted
+pub const NS_TMPFS_LOCATION: &str = "/run/stratisd/ns_mounts";
+
+/// Check if the stratisd mount namespace for this thread is in the root namespace.
+pub fn is_in_root_namespace() -> StratisResult<bool> {
+    let pid_one_stat = stat(INIT_MNT_NS_PATH)?;
+    let self_stat = stat(format!("/proc/self/task/{}/ns/mnt", gettid()).as_str())?;
+    Ok(pid_one_stat.st_ino == self_stat.st_ino)
```

Comment:
Should this also verify that `pid_one_stat.st_dev == self_stat.st_dev`? Man 7 namespaces says this:

```
In  Linux  3.7  and earlier, these files were visible as hard links.  Since Linux 3.8,
they appear as symbolic links.  If two processes are in the same namespace,
then the device IDs and inode numbers of their /proc/[pid]/ns/xxx symbolic links
will be the same; an application can check this using the stat.st_dev and
stat.st_ino fields returned by stat(2). The content of this symbolic link is a string
containing the namespace type and inode number as in the following example:

           $ readlink /proc/$$/ns/uts
           uts:[4026531838]
```

I can't think of a "normal" situation where the `stat.st_dev` of two paths in procfs would differ (all files in procfs report the same device - 0x16/22d) but historically it was possible to bind mount files or directories over parts of `/proc`, which could make this assumption false.

Trying that on a recent kernel appears to fail in all cases I can think of so far, but equally I can't think of a case where checking for an `st_dev` match would cause a problem?

Reply:
I'm fine with adding additional protections! I'll make it check both.